maskit/masks.py

```diff
@@ -0,0 +1,296 @@
+import random as rand
+import pennylane.numpy as np
+from enum import Enum
+from typing import Optional
+
+rand.seed(1337)
+
+
+class PerturbationAxis(Enum):
+    #: Perturbation affects whole wires
+    WIRES = 0
+    #: Perturbation affects whole layers
+    LAYERS = 1
+    #: Perturbation affects random locations in parameter mask
+    RANDOM = 2
+
+
+class PerturbationMode(Enum):
+    #: Adding new holes to the mask
+    ADD = 0
+    #: Removing holes from the mask
+    REMOVE = 1
+    #: Invert current state of the mask
+    INVERT = 2
+
+
+class MaskedObject(object):
+    """
+    A MaskedObject encapsulates a :py:attr:`~.mask` storing boolean value if
+    a specific value is masked or not. In case a specific position is `True`,
+    the according value is masked, otherwise it is not.
+    """
+
+    __slots__ = "_mask"
+
+    def __setitem__(self, key, value: bool):
+        """
+        Convenience function to set the value of a specific position of the
+        encapsulated :py:attr:`~.mask`.
+        """
+        if isinstance(key, int) or isinstance(key, slice) or isinstance(key, tuple):
+            self._mask[key] = value
+        else:
+            raise NotImplementedError
+
+    def __getitem__(self, key):
+        """
+        Convenience function to get the value of a specific position of the
+        encapsulated :py:attr:`~.mask`.
+        """
+        if isinstance(key, int) or isinstance(key, slice) or isinstance(key, tuple):
+            return self._mask[key]
+        raise NotImplementedError
+
+    @property
+    def mask(self):
+        """
+        Returns the encapsulated :py:attr:`~.mask`
+        """
+        return self._mask
+
+    def apply_mask(self, values: np.ndarray):
+        """
+        Applies the encapsulated py:attr:`~.mask` to the given ``values``.
+        Note that the values should have the same shape as the py:attr:`~.mask`.
+
+        :param values: Values where the mask should be applied to
+        """
+        raise NotImplementedError
+
+    def reset(self) -> None:
+        """Resets the mask to not mask anything."""
+        self._mask = np.zeros_like(self._mask, dtype=bool, requires_grad=False)
+
+    def perturb(
+        self,
+        amount: Optional[int] = None,
+        mode: PerturbationMode = PerturbationMode.INVERT,
+    ):
+        """
+        Perturbs the MaskedObject by the given ``mode`` of type
+        :py:class:`~.PerturbationMode` ``amount`` times. If no amount is given,
+        that is ``amount=None``, a random ``amount`` is determined given by the
+        actual size of the py:attr:`~.mask`. The ``amount`` is automatically
+        limited to the actual size of the py:attr:`~.mask`.
+
+        :param amount: Number of items to perturb, defaults to None
+        :param mode: How to perturb, defaults to PerturbationMode.INVERT
+        :raises AssertionError: Raised for negative amounts
```

Comment:
Consider not to document ``AssertionError``, as ``assert`` is skipped when running Python in optimised mode. Rather, document that ``amount`` may not be negative (and omit what happens otherwise).

Reply:
NB: If you *do* want an error, ``ValueError`` seems most appropriate.